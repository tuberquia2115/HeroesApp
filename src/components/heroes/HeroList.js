import React, { useMemo } from 'react'
import { getHeroesByPublisher } from '../../selectors/getHeroesByPublisher'
import { HeroCard } from './HeroCard'


export const HeroList = ({ publisher }) => {
    let heroes = useMemo(() => getHeroesByPublisher(publisher), [publisher])
    return (
        <div className="row align-items-start animate__animated animate__fadeIn">
            {heroes.map((hero) => (
                <HeroCard key={hero.id} {...hero}/>
            ))}

        </div>

    )
}