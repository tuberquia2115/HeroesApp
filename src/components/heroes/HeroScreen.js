import React, { useMemo } from 'react'
import { Redirect, useParams } from 'react-router-dom'
import { getHeroById } from '../../selectors/getHeroById';
export const HeroScreen = ({ history }) => {
    const { heroId } = useParams();
    let hero = useMemo(() => getHeroById(heroId), [heroId])
    console.log(hero)
    if (!hero[0]) {
        return <Redirect to="/" />
    }

    const handlenReturn = () => {
        if (history.length <= 2) {
            history.push('/')
        } else {
            history.goBack()
        }
    }
    const {
        superhero,
        publisher,
        alter_ego,
        first_appearance,
        characters,
    } = hero[0];
    return (
        <div className="row mt-5">
            <hr />
            <div className="col-4">
                <img src={"https://img.pngio.com/iron-man-edwin-jarvis-hulk-falcon-marvel-heroes-2016-cc0-marvel-heroes-png-600_750.jpg"} className="img-thumbnail animate__animated animate__fadeInLeft" alt={superhero} />
            </div>
            <div className="col-8 animate__animated animate__fadeIn">
                <h3 className>{superhero}</h3>
                <ul className="list-group list-group-flush">
                    <li className="list-group-item"> <b> Alter ego: </b>{alter_ego} </li>
                    <li className="list-group-item"> <b> Publisher: </b>{publisher} </li>
                    <li className="list-group-item"> <b> First appearance: </b>{first_appearance} </li>
                </ul>
                <h5>Characters</h5>
                <p>{characters}</p>
                <button
                    onClick={handlenReturn}
                    className="btn btn-outline-info">
                    Return
                </button>
            </div>
        </div>
    )
}