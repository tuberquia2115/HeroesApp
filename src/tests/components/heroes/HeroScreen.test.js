import React from 'react'
import '@testing-library/jest-dom';
import { mount, shallow } from 'enzyme';
import { MemoryRouter, Route } from 'react-router-dom'
import { HeroScreen } from '../../../components/heroes/HeroScreen';

describe('Pruebas en <HeroScreen/>', () => {

    const historyMock = {
        push: jest.fn(),
        length: 10,
        goBack: jest.fn()
    }

    afterEach(() => {
        jest.clearAllMocks()
    })
    test('debe de mostrar el componente redirect si no hau argumentos en el URL', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero']}>
                <HeroScreen
                    history={historyMock}
                />
            </MemoryRouter>
        )
        expect(wrapper.find('Redirect').exists()).toBe(true)
    })
    test('debe de mostrar un hero si el parÃ mentro existe y se encuentra', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route path="/hero/:heroId" component={HeroScreen} />
            </MemoryRouter>
        )

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('img').exists()).toBe(true);
    })

    test('debe de regresar a la pantalla anterior history.push', () => {
        const historyMock = {
            push: jest.fn(),
            length: 1,
            goBack: jest.fn()
        }
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route path="/hero/:heroId" component={() => <HeroScreen history={historyMock} />} />
            </MemoryRouter>
        )

        wrapper.find('button').prop('onClick')();
        expect(historyMock.push).toHaveBeenCalled();
        expect(historyMock.goBack).not.toHaveBeenCalled();
        expect(historyMock.push).toHaveBeenCalledTimes(1)
        expect(historyMock.push).toHaveBeenCalledWith('/')
    })

    test('debe de regresar a la pantalla anterior llamar history.goBack', () => {
        const historyMock = {
            push: jest.fn(),
            length: 4,
            goBack: jest.fn()
        }
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Route path="/hero/:heroId" component={() => <HeroScreen history={historyMock} />} />
            </MemoryRouter>
        )

        wrapper.find('button').prop('onClick')();
        expect(historyMock.goBack).toHaveBeenCalled();
        expect(historyMock.push).not.toHaveBeenCalled();
        expect(historyMock.goBack).toHaveBeenCalledTimes(1)
    })


    test('debe de llamar el redirect si el hero no existe', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider12123']}>
                <Route path="/hero/:heroId" component={() => <HeroScreen history={historyMock} />} />
            </MemoryRouter>
        )
        expect(wrapper.text()).toBe('');

    })



})
