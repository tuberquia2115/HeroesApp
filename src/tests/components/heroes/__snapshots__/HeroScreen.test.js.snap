// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HeroScreen/> debe de mostrar un hero si el par√†mentro existe y se encuentra 1`] = `
<div
  className="row mt-5"
>
  <hr />
  <div
    className="col-4"
  >
    <img
      alt="Spider Man"
      className="img-thumbnail animate__animated animate__fadeInLeft"
      src="https://img.pngio.com/iron-man-edwin-jarvis-hulk-falcon-marvel-heroes-2016-cc0-marvel-heroes-png-600_750.jpg"
    />
  </div>
  <div
    className="col-8 animate__animated animate__fadeIn"
  >
    <h3>
      Spider Man
    </h3>
    <ul
      className="list-group list-group-flush"
    >
      <li
        className="list-group-item"
      >
         
        <b>
           Alter ego: 
        </b>
        Peter Parker
         
      </li>
      <li
        className="list-group-item"
      >
         
        <b>
           Publisher: 
        </b>
        Marvel Comics
         
      </li>
      <li
        className="list-group-item"
      >
         
        <b>
           First appearance: 
        </b>
        Amazing Fantasy #15
         
      </li>
    </ul>
    <h5>
      Characters
    </h5>
    <p>
      Peter Parker
    </p>
    <button
      className="btn btn-outline-info"
      onClick={[Function]}
    >
      Return
    </button>
  </div>
</div>
`;
